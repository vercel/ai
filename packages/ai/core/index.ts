// re-exports:
export { createIdGenerator, generateId } from '@ai-sdk/provider-utils';
export type { IDGenerator } from '@ai-sdk/provider-utils';
export {
  formatAssistantStreamPart,
  formatDataStreamPart,
  jsonSchema,
  parseAssistantStreamPart,
  parseDataStreamPart,
  processDataStream,
  processTextStream,
  zodSchema,
} from '@ai-sdk/ui-utils';
export type {
  AssistantMessage,
  AssistantStatus,
  Attachment,
  ChatRequest,
  ChatRequestOptions,
  CreateMessage,
  DataMessage,
  DataStreamPart,
  DeepPartial,
  IdGenerator,
  JSONValue,
  Message,
  UIMessage,
  RequestOptions,
  Schema,
  ToolInvocation,
  UseAssistantOptions,
} from '@ai-sdk/ui-utils';

// directory exports:
export * from './data-stream';
export * from './embed';
export * from './generate-image';
export * from './generate-object';
export * from './generate-text';
export * from './transcribe';
export * from './middleware';
export * from './prompt';
export * from './registry';
export * from './tool';
export * from './types';

// telemetry types:
export type { TelemetrySettings } from './telemetry/telemetry-settings';

// util exports:
export { cosineSimilarity } from './util/cosine-similarity';
export { simulateReadableStream } from './util/simulate-readable-stream';
