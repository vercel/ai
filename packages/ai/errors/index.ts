export {
  AISDKError,
  APICallError,
  EmptyResponseBodyError,
  InvalidPromptError,
  InvalidResponseDataError,
  JSONParseError,
  LoadAPIKeyError,
  NoContentGeneratedError,
  NoSuchModelError,
  TypeValidationError,
  UnsupportedFunctionalityError,
} from '@ai-sdk/provider';

export { InvalidArgumentError } from './invalid-argument-error';
export { InvalidStreamPartError } from './invalid-stream-part-error';
export { InvalidToolArgumentsError } from './invalid-tool-arguments-error';
export { NoImageGeneratedError } from './no-image-generated-error';
export { NoObjectGeneratedError } from './no-object-generated-error';
export { NoOutputSpecifiedError } from './no-output-specified-error';
export { NoSuchToolError } from './no-such-tool-error';
export { ToolCallRepairError } from './tool-call-repair-error';
export { ToolExecutionError } from './tool-execution-error';
export { MCPClientError } from './mcp-client-error';

export { InvalidDataContentError } from '../core/prompt/invalid-data-content-error';
export { InvalidMessageRoleError } from '../core/prompt/invalid-message-role-error';
export { MessageConversionError } from '../core/prompt/message-conversion-error';
export { DownloadError } from '../util/download-error';
export { RetryError } from '../util/retry-error';
