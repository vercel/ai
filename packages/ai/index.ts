// TODO remove once we can set the source folder in tsconfig.json to src/
export * from './src';
