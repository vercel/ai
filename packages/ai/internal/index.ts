// internal re-exports
export { convertAsyncIteratorToReadableStream } from '@ai-sdk/provider-utils';

// internal
export { convertToLanguageModelPrompt } from '../src/prompt/convert-to-language-model-prompt';
export { prepareToolsAndToolChoice } from '../src/prompt/prepare-tools-and-tool-choice';
export { standardizePrompt } from '../src/prompt/standardize-prompt';
export { prepareCallSettings } from '../src/prompt/prepare-call-settings';
export { prepareRetries } from '../src/util/prepare-retries';
