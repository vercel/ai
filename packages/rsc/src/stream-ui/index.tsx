export { streamUI } from './stream-ui';
