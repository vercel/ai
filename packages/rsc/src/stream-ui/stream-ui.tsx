import { LanguageModelV2, LanguageModelV2CallWarning } from '@ai-sdk/provider';
import { safeParseJSON } from '@ai-sdk/provider-utils';
import { ReactNode } from 'react';
import { z } from 'zod';
import {
  CallWarning,
  FinishReason,
  ProviderOptions,
  LanguageModelUsage,
  ToolChoice,
  Prompt,
  CallSettings,
  InvalidToolArgumentsError,
  NoSuchToolError,
  Schema,
} from 'ai';
import {
  standardizePrompt,
  prepareToolsAndToolChoice,
  prepareRetries,
  prepareCallSettings,
  convertToLanguageModelPrompt,
} from 'ai/internal';
import { createResolvablePromise } from '../util/create-resolvable-promise';
import { isAsyncGenerator } from '../util/is-async-generator';
import { isGenerator } from '../util/is-generator';
import { createStreamableUI } from '../streamable-ui/create-streamable-ui';

type Streamable = ReactNode | Promise<ReactNode>;

type Renderer<T extends Array<any>> = (
  ...args: T
) =>
  | Streamable
  | Generator<Streamable, Streamable, void>
  | AsyncGenerator<Streamable, Streamable, void>;

type RenderTool<PARAMETERS extends z.Schema | Schema = any> = {
  description?: string;
  parameters: PARAMETERS;
  generate?: Renderer<
    [
      PARAMETERS extends z.Schema
        ? z.infer<PARAMETERS>
        : PARAMETERS extends Schema<infer T>
          ? T
          : never,
      {
        toolName: string;
        toolCallId: string;
      },
    ]
  >;
};

type RenderText = Renderer<
  [
    {
      /**
       * The full text content from the model so far.
       */
      content: string;

      /**
       * The new appended text content from the model since the last `text` call.
       */
      delta: string;

      /**
       * Whether the model is done generating text.
       * If `true`, the `content` will be the final output and this call will be the last.
       */
      done: boolean;
    },
  ]
>;

type RenderResult = {
  value: ReactNode;
} & Awaited<ReturnType<LanguageModelV2['doStream']>>;

const defaultTextRenderer: RenderText = ({ content }: { content: string }) =>
  content;

/**
 * `streamUI` is a helper function to create a streamable UI from LLMs.
 */
export async function streamUI<
  TOOLS extends { [name: string]: z.Schema | Schema } = {},
>({
  model,
  tools,
  toolChoice,
  system,
  prompt,
  messages,
  maxRetries,
  abortSignal,
  headers,
  initial,
  text,
  providerOptions,
  onFinish,
  ...settings
}: CallSettings &
  Prompt & {
    /**
     * The language model to use.
     */
    model: LanguageModelV2;

    /**
     * The tools that the model can call. The model needs to support calling tools.
     */
    tools?: {
      [name in keyof TOOLS]: RenderTool<TOOLS[name]>;
    };

    /**
     * The tool choice strategy. Default: 'auto'.
     */
    toolChoice?: ToolChoice<TOOLS>;

    text?: RenderText;
    initial?: ReactNode;

    /**
Additional provider-specific options. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
 */
    providerOptions?: ProviderOptions;

    /**
     * Callback that is called when the LLM response and the final object validation are finished.
     */
    onFinish?: (event: {
      /**
       * The reason why the generation finished.
       */
      finishReason: FinishReason;
      /**
       * The token usage of the generated response.
       */
      usage: LanguageModelUsage;
      /**
       * The final ui node that was generated.
       */
      value: ReactNode;
      /**
       * Warnings from the model provider (e.g. unsupported settings)
       */
      warnings?: CallWarning[];
      /**
       * Optional response data.
       */
      response?: {
        /**
         * Response headers.
         */
        headers?: Record<string, string>;
      };
    }) => Promise<void> | void;
  }): Promise<RenderResult> {
  // TODO: Remove these errors after the experimental phase.
  if (typeof model === 'string') {
    throw new Error(
      '`model` cannot be a string in `streamUI`. Use the actual model instance instead.',
    );
  }
  if ('functions' in settings) {
    throw new Error(
      '`functions` is not supported in `streamUI`, use `tools` instead.',
    );
  }
  if ('provider' in settings) {
    throw new Error(
      '`provider` is no longer needed in `streamUI`. Use `model` instead.',
    );
  }
  if (tools) {
    for (const [name, tool] of Object.entries(tools)) {
      if ('render' in tool) {
        throw new Error(
          'Tool definition in `streamUI` should not have `render` property. Use `generate` instead. Found in tool: ' +
            name,
        );
      }
    }
  }

  const ui = createStreamableUI(initial);

  // The default text renderer just returns the content as string.
  const textRender = text || defaultTextRenderer;

  let finished: Promise<void> | undefined;

  let finishEvent: {
    finishReason: FinishReason;
    usage: LanguageModelUsage;
    warnings?: CallWarning[];
    response?: {
      headers?: Record<string, string>;
    };
  } | null = null;

  async function render({
    args,
    renderer,
    streamableUI,
    isLastCall = false,
  }: {
    renderer: undefined | Renderer<any>;
    args: [payload: any] | [payload: any, options: any];
    streamableUI: ReturnType<typeof createStreamableUI>;
    isLastCall?: boolean;
  }) {
    if (!renderer) return;

    // create a promise that will be resolved when the render call is finished.
    // it is appended to the `finished` promise chain to ensure the render call
    // is finished before the next render call starts.
    const renderFinished = createResolvablePromise<void>();
    finished = finished
      ? finished.then(() => renderFinished.promise)
      : renderFinished.promise;

    const rendererResult = renderer(...args);

    if (isAsyncGenerator(rendererResult) || isGenerator(rendererResult)) {
      while (true) {
        const { done, value } = await rendererResult.next();
        const node = await value;

        if (isLastCall && done) {
          streamableUI.done(node);
        } else {
          streamableUI.update(node);
        }

        if (done) break;
      }
    } else {
      const node = await rendererResult;

      if (isLastCall) {
        streamableUI.done(node);
      } else {
        streamableUI.update(node);
      }
    }

    // resolve the promise to signal that the render call is finished
    renderFinished.resolve(undefined);
  }

  const { retry } = prepareRetries({ maxRetries });

  const validatedPrompt = await standardizePrompt({
    prompt: { system, prompt, messages },
    tools: undefined, // streamUI tools don't support multi-modal tool result conversion
  });
  const result = await retry(async () =>
    model.doStream({
      ...prepareCallSettings(settings),
      ...prepareToolsAndToolChoice({
        tools: tools as any,
        toolChoice,
        activeTools: undefined,
      }),
      prompt: await convertToLanguageModelPrompt({
        prompt: validatedPrompt,
        supportedUrls: await model.supportedUrls,
      }),
      providerOptions,
      abortSignal,
      headers,
    }),
  );

  // For the stream and consume it asynchronously:
  const [stream, forkedStream] = result.stream.tee();
  (async () => {
    try {
      let content = '';
      let hasToolCall = false;
      let warnings: LanguageModelV2CallWarning[] | undefined;

      const reader = forkedStream.getReader();
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        switch (value.type) {
          case 'stream-start': {
            warnings = value.warnings;
            break;
          }

          case 'text': {
            content += value.text;
            render({
              renderer: textRender,
              args: [{ content, done: false, delta: value.text }],
              streamableUI: ui,
            });
            break;
          }

          case 'tool-call-delta': {
            hasToolCall = true;
            break;
          }

          case 'tool-call': {
            const toolName = value.toolName as keyof TOOLS & string;

            if (!tools) {
              throw new NoSuchToolError({ toolName });
            }

            const tool = tools[toolName];
            if (!tool) {
              throw new NoSuchToolError({
                toolName,
                availableTools: Object.keys(tools),
              });
            }

            hasToolCall = true;
            const parseResult = await safeParseJSON({
              text: value.args,
              schema: tool.parameters,
            });

            if (parseResult.success === false) {
              throw new InvalidToolArgumentsError({
                toolName,
                toolArgs: value.args,
                cause: parseResult.error,
              });
            }

            render({
              renderer: tool.generate,
              args: [
                parseResult.value,
                {
                  toolName,
                  toolCallId: value.toolCallId,
                },
              ],
              streamableUI: ui,
              isLastCall: true,
            });

            break;
          }

          case 'error': {
            throw value.error;
          }

          case 'finish': {
            finishEvent = {
              finishReason: value.finishReason,
              usage: value.usage,
              warnings,
              response: result.response,
            };
            break;
          }
        }
      }

      if (!hasToolCall) {
        render({
          renderer: textRender,
          args: [{ content, done: true }],
          streamableUI: ui,
          isLastCall: true,
        });
      }

      await finished;

      if (finishEvent && onFinish) {
        await onFinish({
          ...finishEvent,
          value: ui.value,
        });
      }
    } catch (error) {
      // During the stream rendering, we don't want to throw the error to the
      // parent scope but only let the React's error boundary to catch it.
      ui.error(error);
    }
  })();

  return {
    ...result,
    stream,
    value: ui.value,
  };
}
