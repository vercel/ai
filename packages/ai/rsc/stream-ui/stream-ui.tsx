import { LanguageModelV1 } from '@ai-sdk/provider';
import { safeParseJSON } from '@ai-sdk/provider-utils';
import { ReactNode } from 'react';
import { z } from 'zod';
import { CallSettings } from '../../core/prompt/call-settings';
import { convertToLanguageModelPrompt } from '../../core/prompt/convert-to-language-model-prompt';
import { prepareCallSettings } from '../../core/prompt/prepare-call-settings';
import { prepareRetries } from '../../core/prompt/prepare-retries';
import { prepareToolsAndToolChoice } from '../../core/prompt/prepare-tools-and-tool-choice';
import { Prompt } from '../../core/prompt/prompt';
import { standardizePrompt } from '../../core/prompt/standardize-prompt';
import {
  CallWarning,
  FinishReason,
  ProviderMetadata,
  ToolChoice,
} from '../../core/types';
import { ProviderOptions } from '../../core/types/provider-metadata';
import {
  LanguageModelUsage,
  calculateLanguageModelUsage,
} from '../../core/types/usage';
import { InvalidToolArgumentsError } from '../../errors/invalid-tool-arguments-error';
import { NoSuchToolError } from '../../errors/no-such-tool-error';
import { createResolvablePromise } from '../../util/create-resolvable-promise';
import { isAsyncGenerator } from '../../util/is-async-generator';
import { isGenerator } from '../../util/is-generator';
import { createStreamableUI } from '../streamable-ui/create-streamable-ui';

type Streamable = ReactNode | Promise<ReactNode>;

type Renderer<T extends Array<any>> = (
  ...args: T
) =>
  | Streamable
  | Generator<Streamable, Streamable, void>
  | AsyncGenerator<Streamable, Streamable, void>;

type RenderTool<PARAMETERS extends z.ZodTypeAny = any> = {
  description?: string;
  parameters: PARAMETERS;
  generate?: Renderer<
    [
      z.infer<PARAMETERS>,
      {
        toolName: string;
        toolCallId: string;
      },
    ]
  >;
};

type RenderText = Renderer<
  [
    {
      /**
       * The full text content from the model so far.
       */
      content: string;

      /**
       * The new appended text content from the model since the last `text` call.
       */
      delta: string;

      /**
       * Whether the model is done generating text.
       * If `true`, the `content` will be the final output and this call will be the last.
       */
      done: boolean;

       /**
         * The provider metadata.
         */
       providerMetadata: ProviderMetadata;
    },
  ]
>;

type RenderResult = {
  value: ReactNode;
} & Awaited<ReturnType<LanguageModelV1['doStream']>>;

const defaultTextRenderer: RenderText = ({ content }: { content: string }) =>
  content;

/**
 * `streamUI` is a helper function to create a streamable UI from LLMs.
 */
export async function streamUI<
  TOOLS extends { [name: string]: z.ZodTypeAny } = {},
>({
  model,
  tools,
  toolChoice,
  system,
  prompt,
  messages,
  maxRetries,
  abortSignal,
  headers,
  initial,
  text,
  experimental_providerMetadata,
  providerOptions = experimental_providerMetadata,
  onFinish,
  ...settings
}: CallSettings &
  Prompt & {
    /**
     * The language model to use.
     */
    model: LanguageModelV1;

    /**
     * The tools that the model can call. The model needs to support calling tools.
     */
    tools?: {
      [name in keyof TOOLS]: RenderTool<TOOLS[name]>;
    };

    /**
     * The tool choice strategy. Default: 'auto'.
     */
    toolChoice?: ToolChoice<TOOLS>;

    text?: RenderText;
    initial?: ReactNode;

    /**
Additional provider-specific options. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
 */
    providerOptions?: ProviderOptions;

    /**
@deprecated Use `providerOptions` instead.
*/
    experimental_providerMetadata?: ProviderMetadata;

    /**
     * Callback that is called when the LLM response and the final object validation are finished.
     */
    onFinish?: (event: {
      /**
       * The reason why the generation finished.
       */
      finishReason: FinishReason;
      /**
       * The token usage of the generated response.
       */
      usage: LanguageModelUsage;
      /**
       * The final ui node that was generated.
       */
      value: ReactNode;
      /**
       * Warnings from the model provider (e.g. unsupported settings)
       */
      warnings?: CallWarning[];
      /**
       * Optional raw response data.
       */
      rawResponse?: {
        /**
         * Response headers.
         */
        headers?: Record<string, string>;
      };
    }) => Promise<void> | void;
  }): Promise<RenderResult> {
  // TODO: Remove these errors after the experimental phase.
  if (typeof model === 'string') {
    throw new Error(
      '`model` cannot be a string in `streamUI`. Use the actual model instance instead.',
    );
  }
  if ('functions' in settings) {
    throw new Error(
      '`functions` is not supported in `streamUI`, use `tools` instead.',
    );
  }
  if ('provider' in settings) {
    throw new Error(
      '`provider` is no longer needed in `streamUI`. Use `model` instead.',
    );
  }
  if (tools) {
    for (const [name, tool] of Object.entries(tools)) {
      if ('render' in tool) {
        throw new Error(
          'Tool definition in `streamUI` should not have `render` property. Use `generate` instead. Found in tool: ' +
            name,
        );
      }
    }
  }

  const ui = createStreamableUI(initial);

  // The default text renderer just returns the content as string.
  const textRender = text || defaultTextRenderer;

  let finished: Promise<void> | undefined;

  let finishEvent: {
    finishReason: FinishReason;
    usage: LanguageModelUsage;
    warnings?: CallWarning[];
    rawResponse?: {
      headers?: Record<string, string>;
    };
    providerMetadata: ProviderMetadata;
  } | null = null;

  async function render({
    args,
    renderer,
    streamableUI,
    isLastCall = false,
  }: {
    renderer: undefined | Renderer<any>;
    args: [payload: any] | [payload: any, options: any];
    streamableUI: ReturnType<typeof createStreamableUI>;
    isLastCall?: boolean;
  }) {
    if (!renderer) return;

    // create a promise that will be resolved when the render call is finished.
    // it is appended to the `finished` promise chain to ensure the render call
    // is finished before the next render call starts.
    const renderFinished = createResolvablePromise<void>();
    finished = finished
      ? finished.then(() => renderFinished.promise)
      : renderFinished.promise;

    const rendererResult = renderer(...args);

    if (isAsyncGenerator(rendererResult) || isGenerator(rendererResult)) {
      while (true) {
        const { done, value } = await rendererResult.next();
        const node = await value;

        if (isLastCall && done) {
          streamableUI.done(node);
        } else {
          streamableUI.update(node);
        }

        if (done) break;
      }
    } else {
      const node = await rendererResult;

      if (isLastCall) {
        streamableUI.done(node);
      } else {
        streamableUI.update(node);
      }
    }

    // resolve the promise to signal that the render call is finished
    renderFinished.resolve(undefined);
  }

  const { retry } = prepareRetries({ maxRetries });

  const validatedPrompt = standardizePrompt({
    prompt: { system, prompt, messages },
    tools: undefined, // streamUI tools don't support multi-modal tool result conversion
  });
  const result = await retry(async () =>
    model.doStream({
      mode: {
        type: 'regular',
        ...prepareToolsAndToolChoice({
          tools,
          toolChoice,
          activeTools: undefined,
        }),
      },
      ...prepareCallSettings(settings),
      inputFormat: validatedPrompt.type,
      prompt: await convertToLanguageModelPrompt({
        prompt: validatedPrompt,
        modelSupportsImageUrls: model.supportsImageUrls,
        modelSupportsUrl: model.supportsUrl?.bind(model), // support 'this' context
      }),
      providerMetadata: providerOptions,
      abortSignal,
      headers,
    }),
  );

  // For the stream and consume it asynchronously:
  const [stream, forkedStream] = result.stream.tee();
  (async () => {
    try {
      let content = '';
      let hasToolCall = false;

      const reader = forkedStream.getReader();
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        switch (value.type) {
          case 'text-delta': {
            content += value.textDelta;
            render({
              renderer: textRender,
              args: [{ content, done: false, delta: value.textDelta }],
              streamableUI: ui,
            });
            break;
          }

          case 'tool-call-delta': {
            hasToolCall = true;
            break;
          }

          case 'tool-call': {
            const toolName = value.toolName as keyof TOOLS & string;

            if (!tools) {
              throw new NoSuchToolError({ toolName });
            }

            const tool = tools[toolName];
            if (!tool) {
              throw new NoSuchToolError({
                toolName,
                availableTools: Object.keys(tools),
              });
            }

            hasToolCall = true;
            const parseResult = safeParseJSON({
              text: value.args,
              schema: tool.parameters,
            });

            if (parseResult.success === false) {
              throw new InvalidToolArgumentsError({
                toolName,
                toolArgs: value.args,
                cause: parseResult.error,
              });
            }

            render({
              renderer: tool.generate,
              args: [
                parseResult.value,
                {
                  toolName,
                  toolCallId: value.toolCallId,
                },
              ],
              streamableUI: ui,
              isLastCall: true,
            });

            break;
          }

          case 'error': {
            throw value.error;
          }

          case 'finish': {
            finishEvent = {
              finishReason: value.finishReason,
              usage: calculateLanguageModelUsage(value.usage),
              warnings: result.warnings,
              rawResponse: result.rawResponse,
              providerMetadata: result.providerMetadata,
            };
            break;
          }
        }
      }

      if (!hasToolCall) {
        render({
          renderer: textRender,
          args: [{ content, done: true, providerMetadata: finishEvent?.providerMetadata || {} }],
          streamableUI: ui,
          isLastCall: true,
        });
      }

      await finished;

      if (finishEvent && onFinish) {
        await onFinish({
          ...finishEvent,
          value: ui.value,
        });
      }
    } catch (error) {
      // During the stream rendering, we don't want to throw the error to the
      // parent scope but only let the React's error boundary to catch it.
      ui.error(error);
    }
  })();

  return {
    ...result,
    stream,
    value: ui.value,
  };
}
