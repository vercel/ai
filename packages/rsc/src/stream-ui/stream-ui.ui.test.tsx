import { delay } from '@ai-sdk/provider-utils';
import { convertArrayToReadableStream } from '@ai-sdk/provider-utils/test';
import { MockLanguageModelV2 } from 'ai/test';
import { streamUI } from './stream-ui';
import { z } from 'zod';
import { LanguageModelUsage } from 'ai';

async function recursiveResolve(val: any): Promise<any> {
  if (val && typeof val === 'object' && typeof val.then === 'function') {
    return await recursiveResolve(await val);
  }

  if (Array.isArray(val)) {
    return await Promise.all(val.map(recursiveResolve));
  }

  if (val && typeof val === 'object') {
    const result: any = {};
    for (const key in val) {
      result[key] = await recursiveResolve(val[key]);
    }
    return result;
  }

  return val;
}

async function simulateFlightServerRender(node: React.ReactNode) {
  async function traverse(node: any): Promise<any> {
    if (!node) return {};

    // Let's only do one level of promise resolution here. As it's only for testing purposes.
    const props = await recursiveResolve({ ...node.props });

    const { type } = node;
    const { children, ...otherProps } = props;
    const typeName = typeof type === 'function' ? type.name : String(type);

    return {
      type: typeName,
      props: otherProps,
      children:
        typeof children === 'string'
          ? children
          : Array.isArray(children)
            ? children.map(traverse)
            : await traverse(children),
    };
  }

  return traverse(node);
}

const testUsage: LanguageModelUsage = {
  inputTokens: 3,
  outputTokens: 10,
  totalTokens: 13,
};

const mockTextModel = new MockLanguageModelV2({
  doStream: async () => {
    return {
      stream: convertArrayToReadableStream([
        { type: 'text', text: '{ ' },
        { type: 'text', text: '"content": ' },
        { type: 'text', text: `"Hello, ` },
        { type: 'text', text: `world` },
        { type: 'text', text: `!"` },
        { type: 'text', text: ' }' },
        {
          type: 'finish',
          finishReason: 'stop',
          usage: testUsage,
        },
      ]),
    };
  },
});

const mockToolModel = new MockLanguageModelV2({
  doStream: async () => {
    return {
      stream: convertArrayToReadableStream([
        {
          type: 'tool-call',
          toolCallType: 'function',
          toolCallId: 'call-1',
          toolName: 'tool1',
          args: `{ "value": "value" }`,
        },
        {
          type: 'finish',
          finishReason: 'stop',
          usage: testUsage,
        },
      ]),
    };
  },
});

describe('result.value', () => {
  it('should render text', async () => {
    const result = await streamUI({
      model: mockTextModel,
      prompt: '',
    });

    const rendered = await simulateFlightServerRender(result.value);
    expect(rendered).toMatchSnapshot();
  });

  it('should render text function returned ui', async () => {
    const result = await streamUI({
      model: mockTextModel,
      prompt: '',
      text: ({ content }) => <h1>{content}</h1>,
    });

    const rendered = await simulateFlightServerRender(result.value);
    expect(rendered).toMatchSnapshot();
  });

  it('should render tool call results', async () => {
    const result = await streamUI({
      model: mockToolModel,
      prompt: '',
      tools: {
        tool1: {
          description: 'test tool 1',
          parameters: z.object({
            value: z.string(),
          }),
          generate: async ({ value }) => {
            await delay(100);
            return <div>tool1: {value}</div>;
          },
        },
      },
    });

    const rendered = await simulateFlightServerRender(result.value);
    expect(rendered).toMatchSnapshot();
  });

  it('should render tool call results with generator render function', async () => {
    const result = await streamUI({
      model: mockToolModel,
      prompt: '',
      tools: {
        tool1: {
          description: 'test tool 1',
          parameters: z.object({
            value: z.string(),
          }),
          generate: async function* ({ value }) {
            yield <div>Loading...</div>;
            await delay(100);
            return <div>tool: {value}</div>;
          },
        },
      },
    });

    const rendered = await simulateFlightServerRender(result.value);
    expect(rendered).toMatchSnapshot();
  });

  it('should show better error messages if legacy options are passed', async () => {
    try {
      await streamUI({
        model: mockToolModel,
        prompt: '',
        tools: {
          tool1: {
            description: 'test tool 1',
            parameters: z.object({
              value: z.string(),
            }),
            // @ts-expect-error
            render: async function* () {},
          },
        },
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});

describe('rsc - streamUI() onFinish callback', () => {
  let result: Parameters<
    Required<Parameters<typeof streamUI>[0]>['onFinish']
  >[0];

  beforeEach(async () => {
    const ui = await streamUI({
      model: mockToolModel,
      prompt: '',
      tools: {
        tool1: {
          description: 'test tool 1',
          parameters: z.object({
            value: z.string(),
          }),
          generate: async ({ value }) => {
            await delay(100);
            return <div>tool1: {value}</div>;
          },
        },
      },
      onFinish: event => {
        result = event;
      },
    });

    // consume stream
    await simulateFlightServerRender(ui.value);
  });

  it('should contain token usage', () => {
    expect(result.usage).toStrictEqual(testUsage);
  });

  it('should contain finish reason', async () => {
    expect(result.finishReason).toBe('stop');
  });

  it('should contain final React node', async () => {
    expect(result.value).toMatchSnapshot();
  });
});

describe('options.headers', () => {
  it('should pass headers to model', async () => {
    const result = await streamUI({
      model: new MockLanguageModelV2({
        doStream: async ({ headers }) => {
          expect(headers).toStrictEqual({
            'custom-request-header': 'request-header-value',
          });

          return {
            stream: convertArrayToReadableStream([
              {
                type: 'text',
                text: '{ "content": "headers test" }',
              },
              {
                type: 'finish',
                finishReason: 'stop',
                usage: testUsage,
              },
            ]),
          };
        },
      }),
      prompt: '',
      headers: { 'custom-request-header': 'request-header-value' },
    });

    expect(await simulateFlightServerRender(result.value)).toMatchSnapshot();
  });
});

describe('options.providerMetadata', () => {
  it('should pass provider metadata to model', async () => {
    const result = await streamUI({
      model: new MockLanguageModelV2({
        doStream: async ({ providerOptions }) => {
          expect(providerOptions).toStrictEqual({
            aProvider: { someKey: 'someValue' },
          });

          return {
            stream: convertArrayToReadableStream([
              {
                type: 'text',
                text: '{ "content": "provider metadata test" }',
              },
              {
                type: 'finish',
                finishReason: 'stop',
                usage: testUsage,
              },
            ]),
          };
        },
      }),
      prompt: '',
      providerOptions: {
        aProvider: { someKey: 'someValue' },
      },
    });

    expect(await simulateFlightServerRender(result.value)).toMatchSnapshot();
  });
});
