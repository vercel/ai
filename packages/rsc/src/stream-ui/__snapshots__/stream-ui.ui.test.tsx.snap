// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options.headers > should pass headers to model 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "c": undefined,
      "n": {
        "done": false,
        "next": {
          "done": true,
          "value": "{ "content": "headers test" }",
        },
        "value": "{ "content": "headers test" }",
      },
    },
    "type": "",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`options.providerMetadata > should pass provider metadata to model 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "c": undefined,
      "n": {
        "done": false,
        "next": {
          "done": true,
          "value": "{ "content": "provider metadata test" }",
        },
        "value": "{ "content": "provider metadata test" }",
      },
    },
    "type": "",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`result.value > should render text 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "c": undefined,
      "n": {
        "done": false,
        "next": {
          "done": false,
          "next": {
            "done": false,
            "next": {
              "done": false,
              "next": {
                "done": false,
                "next": {
                  "done": false,
                  "next": {
                    "done": true,
                    "value": "{ "content": "Hello, world!" }",
                  },
                  "value": "{ "content": "Hello, world!" }",
                },
                "value": "{ "content": "Hello, world!"",
              },
              "value": "{ "content": "Hello, world",
            },
            "value": "{ "content": "Hello, ",
          },
          "value": "{ "content": ",
        },
        "value": "{ ",
      },
    },
    "type": "",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`result.value > should render text function returned ui 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "c": undefined,
      "n": {
        "done": false,
        "next": {
          "done": false,
          "next": {
            "done": false,
            "next": {
              "done": false,
              "next": {
                "done": false,
                "next": {
                  "done": false,
                  "next": {
                    "done": true,
                    "value": <h1>
                      { "content": "Hello, world!" }
                    </h1>,
                  },
                  "value": <h1>
                    { "content": "Hello, world!" }
                  </h1>,
                },
                "value": <h1>
                  { "content": "Hello, world!"
                </h1>,
              },
              "value": <h1>
                { "content": "Hello, world
              </h1>,
            },
            "value": <h1>
              { "content": "Hello, 
            </h1>,
          },
          "value": <h1>
            { "content": 
          </h1>,
        },
        "value": <h1>
          { 
        </h1>,
      },
    },
    "type": "",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`result.value > should render tool call results 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "c": undefined,
      "n": {
        "done": true,
        "value": <div>
          tool1: 
          value
        </div>,
      },
    },
    "type": "",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`result.value > should render tool call results with generator render function 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "c": undefined,
      "n": {
        "done": false,
        "next": {
          "done": true,
          "value": <div>
            tool: 
            value
          </div>,
        },
        "value": <div>
          Loading...
        </div>,
      },
    },
    "type": "",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`result.value > should show better error messages if legacy options are passed 1`] = `[Error: Tool definition in \`streamUI\` should not have \`render\` property. Use \`generate\` instead. Found in tool: tool1]`;

exports[`rsc - streamUI() onFinish callback > should contain final React node 1`] = `
<React.Suspense>
  <Unknown
    n={Promise {}}
  />
</React.Suspense>
`;
